libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1312,9 +1312,13 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 		}
 		else
 		{
-			IRVariable argument = convert(*arguments.front(), *arrayType.baseType());
+			IRVariable argument =
+				arrayType.baseType()->isValueType() ?
+				convert(*arguments.front(), *arrayType.baseType()) :
+				*arguments.front();
```

Comment:
I wonder whether we could skip this here in the value type case (i.e. never convert here, but always pass ``arraType.baseType()`` to the ``storageArrayPushFunction`` below and instead just convert inside ``storageArrayPushFunction`` (getting rid of the ``solUnimplementedAssert(*_fromType == *_type.baseType(), "");`` in the beginning). But even if so, maybe we don't need to in this PR.

Reply:
Good idea!

Reply:
Added a point to https://github.com/ethereum/solidity/issues/10349 If we fix it there, it will be fixed here, I think.